lib/Dialect/Pulse/Utils/Utils.cpp

```diff
@@ -35,4 +44,91 @@ Waveform_CreateOp getWaveformOp(PlayOp pulsePlayOp,
   return wfrOp;
 }
 
+Waveform_CreateOp getWaveformOp(PlayOp pulsePlayOp,
+                                CallSequenceStack_t &callSequenceOpStack) {
+  auto wfrIndex = 0;
+  mlir::Value wfrOp = pulsePlayOp.getWfr();
+
+  for (auto it = callSequenceOpStack.rbegin(); it != callSequenceOpStack.rend();
+       ++it) {
+    if (wfrOp.isa<BlockArgument>()) {
+      wfrIndex = wfrOp.dyn_cast<BlockArgument>().getArgNumber();
+      wfrOp = it->getOperand(wfrIndex);
+    } else
+      break;
+  }
+
+  auto waveformOp = dyn_cast<Waveform_CreateOp>(wfrOp.getDefiningOp());
+  if (!waveformOp)
+    pulsePlayOp->emitError() << "The wfr argument is not a Waveform_CreateOp.";
+  return waveformOp;
+}
+
+double getPhaseValue(ShiftPhaseOp shiftPhaseOp,
+                     CallSequenceStack_t &callSequenceOpStack) {
+  auto phaseOffsetIndex = 0;
+  mlir::Value phaseOffset = shiftPhaseOp.getPhaseOffset();
+
+  for (auto it = callSequenceOpStack.rbegin(); it != callSequenceOpStack.rend();
+       ++it) {
+    if (phaseOffset.isa<BlockArgument>()) {
+      phaseOffsetIndex = phaseOffset.dyn_cast<BlockArgument>().getArgNumber();
+      phaseOffset = it->getOperand(phaseOffsetIndex);
+    } else
+      break;
+  }
+
+  auto phaseOffsetOp =
+      dyn_cast<mlir::arith::ConstantFloatOp>(phaseOffset.getDefiningOp());
+  if (!phaseOffsetOp)
+    phaseOffsetOp->emitError() << "Phase offset is not a ConstantFloatOp.";
+  return phaseOffsetOp.value().convertToDouble();
+}
+
+void sortOpsByTimepoint(SequenceOp &sequenceOp) {
```

Comment:
I think it would be good to add a doxygen comment for this method (in the header file) explaining exactly how this sort works (i.e., it looks like it is sorting blocks within a region).